livekit-rtc/livekit/rtc/participant.py

```diff
@@ -236,6 +241,183 @@ async def publish_transcription(self, transcription: Transcription) -> None:
+        - `request_id`: A unique identifier for this RPC request
+        - `caller_identity`: The identity of the RemoteParticipant who initiated the RPC call
+        - `payload`: The data sent by the caller (as a string)
+        - `response_timeout_ms`: The maximum time available to return a response
+
+        The handler should return a string or a coroutine that resolves to a string.
+        If unable to respond within `response_timeout_ms`, the request will result in an error on the caller's side.
+
+        You may raise errors of type `RpcError` with a string `message` in the handler,
+        and they will be received on the caller's side with the message intact.
+        Other errors raised in your handler will not be transmitted as-is, and will instead arrive to the caller as `1500` ("Application Error").
+        """
+        self._rpc_handlers[method] = handler
+
+        req = proto_ffi.FfiRequest()
+        req.register_rpc_method.local_participant_handle = self._ffi_handle.handle
+        req.register_rpc_method.method = method
+
+        FfiClient.instance.request(req)
+
+    def rpc_method(self, method: str):
+        """
+        Decorator form of `register_rpc_method`
+
+        Args:
+            method (str): The name of the indicated RPC method
+
+        Example:
+            @local_participant.rpc_method("greet")
+            async def greet_handler(request_id: str, caller_identity: str, payload: str, response_timeout_ms: int) -> str:
+                print(f"Received greeting from {caller_identity}: {payload}")
+                return f"Hello, {caller_identity}!"
+
+        See Also:
+            `register_rpc_method` for more details
+        """
+
+        def decorator(handler: Callable[[str, str, str, int], Awaitable[str]]):
+            self.register_rpc_method(method, handler)
+            return handler
+
+        return decorator
+
+    def unregister_rpc_method(self, method: str) -> None:
+        """
+        Unregisters a previously registered RPC method.
+
+        Args:
+            method (str): The name of the RPC method to unregister
+        """
+        self._rpc_handlers.pop(method, None)
+
+        req = proto_ffi.FfiRequest()
+        req.unregister_rpc_method.local_participant_handle = self._ffi_handle.handle
+        req.unregister_rpc_method.method = method
+
+        FfiClient.instance.request(req)
+
+    async def _handle_rpc_method_invocation(
+        self,
+        invocation_id: int,
+        method: str,
+        request_id: str,
+        caller_identity: str,
+        payload: str,
+        response_timeout_ms: int,
+    ) -> None:
+        response_error: Optional[RpcError] = None
+        response_payload: Optional[str] = None
+
+        handler = self._rpc_handlers.get(method)
+
+        if not handler:
+            response_error = RpcError._built_in(RpcError.ErrorCode.UNSUPPORTED_METHOD)
+        else:
+            try:
+                response_payload = await handler(
```

Comment:
Should we allow non-async handlers too?

Reply:
yeah I think its reasonable. I changed the handler return type to `Union[Awaitable[str], str]` and then in this spot i checked it with `if asyncio.iscoroutinefunction(handler):` before awaiting.